### ISSUE TITLE
Unexpected Client Connection Pool shutdown when connected to FileSink

### ISSUE THREAD

**Issue by ANorwell** [akka/akka-http#552]
Using an akka-http client to stream the body to a file (using FileIO.toPath), I sometimes get a sequence of errors like the following:
```
[ERROR] [17:11:16.937UTC] [lego-akka.actor.default-dispatcher-31] [a.h.i.engine.client.PoolMasterActor] [akka://lego/user/pool-master] - connection pool for PoolGateway(hcps = HostConnectionPoolSetup(host.example.com,443,ConnectionPoolSetup(ConnectionPoolSettings(196,0,0,4096,1,30 seconds,ClientConnectionSettings(Some(User-Agent: akka-http/2.4.11),10 seconds,1 minute,512,<function0>,List(),ParserSettings(2048,16,64,64,8192,64,9223372036854775807,256,10485760,Strict,RFC6265,true,Full,Error,Map(If-Range -> 0, If-Modified-Since -> 0, If-Unmodified-Since -> 0, default -> 12, Content-MD5 -> 0, Date -> 0, If-Match -> 0, If-None-Match -> 0, User-Agent -> 32),false,<function1>,<function1>,<function2>))),akka.http.scaladsl.HttpsConnectionContext@366f4bd2,akka.event.BusLogging@57d25971))) has shut down unexpectedly
> [ERROR] [17:11:17.017UTC] [lego-akka.actor.default-dispatcher-87] [akka.stream.impl.io.FileSubscriber] [akka://lego/user/StreamSupervisor-0/flow-401634-1-fileSink] - Tearing down FileSink(/my/path/file/blob) due to upstream error
> [ERROR] [17:11:17.017UTC] [lego-akka.actor.default-dispatcher-87] [akka.actor.ActorSystemImpl] [akka.actor.ActorSystemImpl(lego)] - File Write error java.util.concurrent.TimeoutException: No elements passed in the last 1 minute.
```

The shutdown of the pool is especially problematic since all outstanding requests being handled by the pool are lost. This only happens to me under load.

Simplified reproduction code would look something like the following:

```scala
object Repr {
  import akka.actor.ActorSystem
  import akka.stream.{ IOResult, ActorMaterializer }
  import akka.stream.scaladsl._
  import akka.util.ByteString
  import akka.http.scaladsl._
  import akka.http.scaladsl.model._
  import java.nio.file._
  import java.nio.file.StandardOpenOption._
  import scala.concurrent.Future

  implicit val system = ActorSystem()
  implicit val mat = ActorMaterializer()

  def requestHandler(req: HttpRequest) = Http().singleRequest(req)

  def writeSource(filePath: Path, source: Source[ByteString, Any]): Future[IOResult] =
    source.runWith(FileIO.toPath(filePath, Set(CREATE, WRITE, TRUNCATE_EXISTING)))

  // running this flow under load would cause the above error
  //Assume that the outer flow does not fail under load (e.g., wrap futures with `Try`)
  val flow = Flow[HttpRequest]
    .mapAsyncUnordered(4)(requestHandler)
    .mapAsyncUnordered(4)(rsp ⇒ writeSource(path(rsp), rsp.entity.dataBytes))

  def path(rsp: HttpResponse) = Paths.get(s"/tmp/${java.util.UUID.randomUUID}")
}
```

**Comment by ktoso** [comment]
Thanks for reporting, the error logged is an idle-timeout (you could disable it in config), but it also means that no data was flowing through the given tcp connection for a minute... Are you sure the writing is able to keep up etc?

**Comment by ANorwell** [comment]
I'll try increasing `idle-timeout`, and investigating whether the disk write can keep up. Thanks for the suggestions.

I wonder, is the connection truly idle here, i.e. server has finished sending packets and connection is idle (at the OS level) for the last 60 seconds? Or does it just appear idle because of back-pressure, and the server is still sending packets that are buffering at the OS level or somewhere else?

**Comment by jrudolph** [comment]
Hmm, the "connection pool for ... has shut down unexpectedly" error message is always a bug. Just to confirm: @ANorwell which akka-http version are you running?

**Comment by ANorwell** [comment]
@jrudolph this is from 2.4.11. I haven't rerun with 10.0.0 under the same conditions, but I'll try to as soon as I get a chance.

**Comment by jrudolph** [comment]
We fixed that a single connection can crash whole pools in #481 which was released with 10.0.0 so it would indeed be interesting to see if that would have fixed your issue.

**Comment by ANorwell** [comment]
I have not seen this on 10.x despite encountering the `No elements passed in the last 1 minute` error, so I assume it was fixed in #481.

**Comment by jrudolph** [comment]
Great, thanks for checking, @ANorwell.
